File: clients/chat-room/docker/nodejs/src/consumer.ts
import { Kafka } from "kafkajs";
import { v4 as uuidv4 } from "uuid";
const redpanda = new Kafka({
  brokers: ["localhost:19092"],
});
const consumer = redpanda.consumer({ groupId: uuidv4() });
export async function connect() {
  try {
    await consumer.connect();
    await consumer.subscribe({ topic: "chat-room" });
    await consumer.run({
      eachMessage: async ({ topic, partition, message }) => {
        const formattedValue = JSON.parse(
          (message.value as Buffer).toString()
        );
        console.log(`${formattedValue.user}: ${formattedValue.message}`);
      },
    });
  } catch (error) {
    console.error("Error:", error);
  }
}
export async function disconnect() {
  try {
    await consumer.disconnect();
  } catch (error) {
    console.error("Error:", error);
  }
}

File: clients/chat-room/docker/go/consumer.go
package main
import (
    "context"
    "encoding/json"
    "fmt"
    "github.com/twmb/franz-go/pkg/kgo"
    "github.com/google/uuid"
)
type Consumer struct {
    client *kgo.Client
    topic  string
}
func NewConsumer(brokers []string, topic string) *Consumer {
    groupID := uuid.New().String()
    client, err := kgo.NewClient(
        kgo.SeedBrokers(brokers...),
        kgo.ConsumerGroup(groupID),
        kgo.ConsumeTopics(topic),
        kgo.ConsumeResetOffset(kgo.NewOffset().AtStart()),
    )
    if err != nil {
        panic(err)
    }
    return &Consumer{client: client, topic: topic}
}
func (c *Consumer) PrintMessages() {
    ctx := context.Background()
    for {
        fetches := c.client.PollFetches(ctx)
        iter := fetches.RecordIter()
        for !iter.Done() {
            record := iter.Next()
            var msg Message
            if err := json.Unmarshal(record.Value, &msg); err != nil {
                fmt.Printf("Error decoding message: %v\n", err)
                continue
            }
            fmt.Printf("%s: %s\n", msg.User, msg.Message)
        }
    }
}
func (c *Consumer) Close() {
    c.client.Close()
}